src/lib/mongoBackend/mongoConnectionPool.cpp

```diff
@@ -0,0 +1,445 @@
+  }
+
+  //
+  // Set up the semaphore protecting the set of connections of the pool (connectionSem)
+  // Note that this is a counting semaphore, initialized to connectionPoolSize.
+  //
+  r = sem_init(&connectionSem, 0, connectionPoolSize);
+  if (r != 0)
+  {
+    LM_E(("Runtime Error (cannot create connection semaphore-set)"));
+    return -1;
+  }
+
+  // Measure accumulated semaphore waiting time?
+  semStatistics = semTimeStat;
+  
+  return 0;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoPoolConnectionGet - 
+*
+* There are two semaphores to get a connection.
+* - One binary semaphore that protects the connection-vector itself (connectionPoolSem)
+* - One counting semaphore that makes the caller wait until there is at least one free connection (connectionSem)
+*
+* There is a limited number of connections and the first thing to do is to wait for a connection
+* to become avilable (any of the N connections in the pool) - this is done waiting on the counting semaphore that is 
+* initialized with "POOL SIZE" - meaning the semaphore can be taken N times if the pool size is N.
+* 
+* Once a connection is free, 'sem_wait(&connectionSem)' returns and we now have to take the semaphore 
+* that protects for pool itself (we *are* going to modify the vector of the pool - can only do it in
+* one thread at a time ...)
+*
+* After taking a connection, the semaphore 'connectionPoolSem' is freed, as all modifications to the connection pool
+* have finished.
+* The other semaphore however, 'connectionSem', is kept and it is not freed until we finish using the connection.
+*
+* The function mongoPoolConnectionRelease releases the counting semaphore 'connectionSem'.
+* Very important to call the function 'mongoPoolConnectionRelease' after finishing using the connection !
+*
+*/
+DBClientBase* mongoPoolConnectionGet(void)
+{
+  DBClientBase*    connection = NULL;
+  struct timespec  startTime;
+  struct timespec  endTime;
+  struct timespec  diffTime;
+
+  if (semStatistics)
+  {
+    clock_gettime(CLOCK_REALTIME, &startTime);
+  }
+
+  sem_wait(&connectionSem);
+  sem_wait(&connectionPoolSem);
+
+  if (semStatistics)
+  {
+    clock_gettime(CLOCK_REALTIME, &endTime);
+
+    clock_difftime(&endTime, &startTime, &diffTime);
+    clock_addtime(&semWaitingTime, &diffTime);
+  }
+
+  for (int ix = 0; ix < connectionPoolSize; ++ix)
+  {
+    if (connectionPool[ix].free == true)
+    {
+      connectionPool[ix].free = false;
+      connection = connectionPool[ix].connection;
+      break;
+    }
+  }
+
+  sem_post(&connectionPoolSem);
+  
+  return connection;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoPoolConnectionRelease - 
+*/
+void mongoPoolConnectionRelease(DBClientBase* connection)
+{
+  sem_wait(&connectionPoolSem);
+
+  for (int ix = 0; ix < connectionPoolSize; ++ix)
+  {
+    if (connectionPool[ix].connection == connection)
```

Comment:
DBClientBase luckily doesn't override the '==' operator.
Not sure I understand your doubt here.

mongoPoolConnectionGet() returns a connection pointer and this very pointer is later sent to mongoPoolConnectionRelease to release it.